File: 2022/typescript/src/day03.ts
import { chunk, intersect, sumOf } from "std/collections/mod.ts";

function run(input: string): void {
    const res1 = part1(input);
    console.log('Day 03, part 1:', res1);
    const res2 = part2(input);
    console.log('Day 03, part 2:', res2);
}

function part1(input: string): number {
    const lines = input.split('\n');

    const splitLines = lines.map(line => {
        const n = line.length / 2;
        const f = line.slice(0, n);
        const s = line.slice(n);
        return [f, s];
    });

    const overlaps = splitLines.map(([f, s]) => {
        const overlap = intersect(Array.from(f), Array.from(s));
        return overlap;
    });

    const priorities = overlaps.map(overlap => {
        return sumOf(overlap, priority)
    });

    const sum = sumOf(priorities, Number);

    return sum;
}

function priority(char: string): number {
    const code = char.charCodeAt(0);
    return (code >= 97) ? code - 96 : code - 38;
}

function part2(input: string): number {
    const lines = input.split('\n');
    // const groups = elfGroups(lines);
    const groups = chunk(lines, 3);

    const overlaps = groups.map((strings) => {
        const as = strings.map(s => s ? Array.from(s) : []);
        const overlap = intersect(...as);
        return overlap;
    });

    const priorities = overlaps.map(overlap => {
        return sumOf(Array.from(overlap), priority)
    });

    const sum = sumOf(priorities, Number);

    return sum;
}

export default { run, part1, part2 };

File: 2022/fsharp/days/Day03.fs
module Day03

let priority (char: char) =
    let code = char |> int
    if code >= 97 then code - 96 else code - 38

let part1 (input: string) =
    let lines = input.Split("\n")

    let splitLines =
        lines
        |> Array.map (fun line -> line |> Seq.toList |> List.splitAt (line.Length / 2))

    let freqSets =
        splitLines |> Array.map (fun (f, s) -> (f |> Set.ofList, s |> Set.ofList))

    let overlaps =
        freqSets
        |> Array.map (fun (f, s) -> Set.intersect f s |> Set.toList)
        |> List.concat

    let sums = overlaps |> List.map priority
    sums |> List.sum

let part2 (input: string) =
    let lines = input.Trim().Split("\n")
    let groups = lines |> Array.chunkBySize 3

    let freqSets =
        groups
        |> Array.map (fun group ->
            group
            |> Array.map (fun line -> line.ToCharArray() |> Array.toList |> Set.ofList))

    let overlaps =
        freqSets
        |> Array.map (fun group ->
            let f = group.[0]
            let s = group.[1]
            let t = group.[2]
            Set.intersect f s |> Set.intersect t |> Set.toList)
        |> List.concat

    let sums = overlaps |> List.map priority
    sums |> List.sum

let run (input: string) =
    printfn "Day 03, part1: %A" (part1 input)
    printfn "Day 03, part2: %A" (part2 input)
